processing/src/main/java/org/apache/druid/query/expression/IPv4AddressMatchExprMacro.java

```diff
@@ -0,0 +1,198 @@
+ * - boolean ipv4address_match(string address, string subnet)
+ * - boolean ipv4address_match(string address, string subnet, boolean inclusive)
+ * - boolean ipv4address_match(long address, string subnet)
+ * - boolean ipv4address_match(long address, string subnet, boolean inclusive)
+ *
+ * Valid "address" argument formats are:
+ * - unsigned int long (e.g., 3232235521)
+ * - unsigned int string (e.g., "3232235521")
+ * - IPv4 address dotted-decimal notation string (e.g., "198.168.0.1")
+ * - IPv6 IPv4-mapped address string (e.g., "::ffff:192.168.0.1")
+ *
+ * The argument format for the "subnet" argument should be a literal in CIDR notation
+ * (e.g., "198.168.0.0/16").
+ *
+ * An optional "inclusive" argument should be a boolean literal (e.g., 1, 0, "true", or "false") that
+ * indicates whether the network and the broadcast addresses should be considered part of the
+ * subnet. When this argument is absent, its value defaults to "false".
+ *
+ * The overloaded signature allows applying the expression to a dimension with mixed string and long
+ * representations of IPv4 addresses.
+ * </pre>
+ *
+ * @see IPv4AddressParseExprMacro
+ * @see IPv4AddressStringifyExprMacro
+ */
+public class IPv4AddressMatchExprMacro implements ExprMacroTable.ExprMacro
+{
+  public static final String NAME = "ipv4address_match";
+  private static final int ARG_SUBNET = 1;
+  private static final int ARG_INCLUSIVE = 2;
+
+  @Override
+  public String name()
+  {
+    return NAME;
+  }
+
+  @Override
+  public Expr apply(final List<Expr> args)
+  {
+    if (args.size() < 2 || 3 < args.size()) {
+      throw new IAE(createErrMsg("must have 2-3 arguments"));
+    }
+
+    boolean inclusive = getInclusive(args);
+    SubnetUtils.SubnetInfo subnetInfo = getSubnetInfo(args, inclusive);
+
+    Expr arg = args.get(0);
+
+    class IPv4AddressMatchExpr extends ExprMacroTable.BaseScalarUnivariateMacroFunctionExpr
+    {
+      private final SubnetUtils.SubnetInfo subnetInfo;
+
+      private IPv4AddressMatchExpr(Expr arg, SubnetUtils.SubnetInfo subnetInfo)
+      {
+        super(arg);
+        this.subnetInfo = subnetInfo;
+      }
+
+      @Nonnull
+      @Override
+      public ExprEval eval(final ObjectBinding bindings)
+      {
+        ExprEval eval = arg.eval(bindings);
+        boolean match;
+        switch (eval.type()) {
+          case STRING:
+            match = isStringMatch(eval.asString());
+            break;
+          case LONG:
+            match = !eval.isNumericNull() && isLongMatch(eval.asLong());
+            break;
+          default:
+            match = false;
+        }
+        return ExprEval.of(match, ExprType.LONG);
+      }
+
+      private boolean isStringMatch(String stringValue)
+      {
+        boolean match;
+        String ipv4 = IPv4AddressExprUtils.extractIPv4Address(stringValue);
+        if (ipv4 == null) {
+          match = isLongMatch(stringValue);
+        } else {
+          match = subnetInfo.isInRange(ipv4);
```

Comment:
It looks like `SubnetInfo.isInRange(String)` converts the String address into an int anyway. We could have fed it an int directly. May be nicer to do something like

```java
subnetInfo.isInRange(
  IPv4AddressExprUtils.inet4AddressToInt(
    IPv4AddressExprUtils.extractIPv4Address(stringValue)
  )
)
```

(With the appropriate null checks sprinkled in)